sprockets/mixins/mediatype/type_info.py

```diff
@@ -0,0 +1,110 @@
+from __future__ import annotations
+
+import typing
+import uuid
+
+try:
+    from typing import Protocol
+except ImportError:
+    # "ignore" is required to avoid an incompatible import
+    # error due to different bindings of _SpecialForm
+    from typing_extensions import Protocol  # type: ignore
+
+
+class DefinesIsoFormat(Protocol):
+    """An object that has an isoformat method."""
+    def isoformat(self) -> str:
+        """Return the date/time in ISO-8601 format."""
+        ...
+
+
+class HasSettings(Protocol):
+    """Something that quacks like a tornado.web.Application."""
+    settings: typing.Dict[str, typing.Any]
+    """Application settings."""
+
+
+Serializable = typing.Union[DefinesIsoFormat, None, bool, bytearray, bytes,
+                            float, int, memoryview, str, typing.Mapping,
+                            typing.Sequence, typing.Set, uuid.UUID]
```

Comment:
Should the collection types be recursive, e.g. `typing.Sequence['Serializable']`? I see there's an open issue for this in mypy, but they appear to support recursive types in some circumstances: https://github.com/python/mypy/issues/731#issuecomment-922760064

Reply:
Yeah ... recursive types in this case anger mypy:
```
sprockets/mixins/mediatype/type_info.py:27: error: Cannot resolve name "Serializable" (possible cyclic definition)
sprockets/mixins/mediatype/type_info.py:29: error: Cannot resolve name "Serializable" (possible cyclic definition)
sprockets/mixins/mediatype/type_info.py:30: error: Cannot resolve name "Serializable" (possible cyclic definition)
sprockets/mixins/mediatype/type_info.py:49: error: Cannot resolve name "PackBFunction" (possible cyclic definition)
sprockets/mixins/mediatype/type_info.py:55: error: Cannot resolve name "DumpSFunction" (possible cyclic definition)
Found 5 errors in 1 file (checked 8 source files)
```

I've been tinkering with using [pyright](https://github.com/microsoft/pyright) as well since it can handle things like:
```python
Serializable = typing.Union[DefinesIsoFormat, None, bool, bytearray, bytes,
                            float, int, memoryview, str,
                            typing.Mapping[str, 'Serializable'],
                            typing.Sequence['Serializable'],
                            typing.Set['Serializable'],
                            uuid.UUID]
```

However, it's Python parser does not understand things like :(
```python
try:
    from typing import Protocol
except ImportError:
    # "ignore" is required to avoid an incompatible import
    # error due to different bindings of _SpecialForm
    from typing_extensions import Protocol  # type: ignore
```

Anyway .... that is a really long way of saying, _let's revisit this once we find a pattern that works ;) _

Reply:
Always something! :)

Sounds good to me